### PR TITLE
feat(ui): Notifications and Popups

### PR DESCRIPTION
**Pull request by 2hwk**
Fixes #[issue_no]

## Summary of Changes

1. Notify the user when a setting change is made that requires an airplane reload. 
- Creates a Pop-Up that takes focus of the user's screen until dismissed (Only on MCDU)
- Creates a notification which persists for 10 seconds.

2. Notify the user to calibrate their throttles in the EFB
- Creates a notification that persists for 1 minute when the following conditions have been met:
  - Calibration has not been completed in the EFB
  - At least 10 minutes have passed
  - Physical throttle 1 axis has not been moved in that time
  - Flight is in a cruise phase (as determined by FWC)
  - LVR CLB message is showing on the PFD.
  - Throttle 1 is at a position above a point slightly below the lower bounds of the climb detent.

## Screenshots (if necessary)

https://user-images.githubusercontent.com/15316958/124350034-ceb78880-dc24-11eb-9773-28725a9ad143.mp4

Note: Notification styles have changed since WU5. The above video is outdated.

https://cdn.discordapp.com/attachments/747571985831297145/874942900390223872/nh0B9U0bCq.webm

## References

N/A, sim UX related PR.

## Additional context

Also added utility clases to A32NX_Utils.js (for JS) and @instruments/common (notification/popup.ts) for posterity.

## Code Examples

**React**
```
import { PopUp } from '@instruments/common/popup';
import { Notification } from '@instruments/common/notification';

...
const popup = new PopUp();
const notification = new Notification();
const yesFunc = () => {
    notification.showNotification({ message: 'Your notification here!' });
};
popup.showPopUp({ title: "YOUR TITLE HERE", message: "Your message here", style: "small"}, yesFunc, noFunc);
...
```

**HTML+JS**
```
const popup = new NXPopUp();
const notif = new NXNotif();
const yesFunc = () => {
    notif.showNotification({ message:"Your message here"});
};
popup.showPopUp({ title: "YOUR TITLE HERE", message: "Your message here", style: "small"}, yesFunc, noFunc);
```

Discord username (if different from GitHub): 2Cas#1022

## Testing instructions

**Please note if you are using the MS Store version or Steam version**

For settings reload notification:

1. Load the plane in any state
2. Change a MCDU option that requires a plane reload (KG/LB conversion or Keyboard Input)
3. Observe Pop-Up with Y/N options
4. When Pop-Up is cleared, a notification (w associated sound event) will display for 10 seconds on the HUD in the top center of the screen.

For calibration notification:

5. Delete any previous throttle calibration. Open File Explorer, go to %appdata%/Microsoft Flight Simulator/Packages/flybywire-aircraft-a320-neo/work and delete ThrottleConfiguration.ini. (N.B. For convenience, you don't actually need to delete this file, just move it out of the folder and back it up somewhere safe). 
6. Go to cruise (CRZ page on lower ECAM), making sure that AP is on and A/THRST is active. Set throttle either on the CLB detent, or a bit behind it, such that LVR CLB flashes on the PFD. Leave it in this state for at least 10 minutes, do not move the throttles.
7. A notification should pop up, and this will persist for a minute before disappearing. This notification will reappear every 5 minutes if the problem still exists.
8.  Enter the EFB and calibrate as per normal. On the next flight, the notification will not trigger even if the above steps are repeated, as long as a custom calibration has been set.

Known Issues:

* English only, no localisation possible for now.
* Default sim's Notifications can clash if enabled in the assistants options, and this behavior can be jarring at times.

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Comment by 2hwk** [comment]
I have a feeling SU5 might have some major under the hood changes, so any progress on this will be delayed until after then.

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_FMGC.js [hunk 1]
```diff
@@ -102,6 +102,12 @@ class CDU_OPTIONS_FMGC {
             } else {
                 NXDataStore.set("CONFIG_USING_METRIC_UNIT", storedUsingMetric === 1 ? "2.20462" : "1");
                 mcdu.addNewMessage(NXFictionalMessages.reloadPlaneApply);
+                const notif = new NXNotif();
+                const popup = new NXPopUp();
+                const showNotif = () => {
+                    notif.showNotification({title: "RELOAD AIRCRAFT", theme: "GAMEPLAY", message:"Reload the aircraft to apply settings"});
```
Comment (Benjozork) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_FMGC.js:108]:
```suggestion
                    notif.showNotification({ title: "RELOAD AIRCRAFT", theme: "GAMEPLAY", message:"Reload the aircraft to apply settings" });
```

File: src/instruments/src/Common/notification.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
+/* eslint-disable camelcase */
```
Comment (Benjozork) [inline src/instruments/src/Common/notification.ts:1]: **suggestion:** move those files to `src/shared/`, as they might be used by non-instrument components such as the FMGC.

File: src/instruments/src/Common/popup.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
+/* eslint-disable no-undef */
```
Comment (Benjozork) [inline src/instruments/src/Common/popup.ts:1]: **suggestion:** move those files to `src/shared/`, as they might be used by non-instrument components such as the FMGC.

File: src/instruments/src/Common/notification.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    }
+}
+
+/**
+Usage:
+import { Notification } from '@instruments/common/notification';
+
+const notification = new Notification();
+notification.showNotification({ message: 'Your notification here!' });
+ */
```
Comment (Benjozork) [inline src/instruments/src/Common/notification.ts:90]: **issue:** This doc comment should be on the relevant element, in this case the `Notification` class.

File: src/instruments/src/Common/popup.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    }
+}
+
+/**
+Usage:
+import { PopUp } from '@instruments/common/popup';
+
+const popup = new PopUp();
+popup.showPopUp({ title: "CRITICAL SETTING CHANGED", message: "Your message here", style: "small"}, yesFunc, noFunc);
+*/
```
Comment (Benjozork) [inline src/instruments/src/Common/popup.ts:92]: **issue:** This doc comment should be on the relevant element, in this case the `PopUp` class.

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_MCDU_KB.js [hunk 6]
```diff
@@ -32,6 +32,12 @@ class CDU_OPTIONS_MCDU_KB {
                 mcdu.clearFocus();
                 NXDataStore.set("MCDU_KB_INPUT", "ENABLED");
                 mcdu.addNewMessage(NXFictionalMessages.reloadPlaneApply);
+                const notif = new NXNotif();
+                const popup = new NXPopUp();
+                const showNotif = () => {
+                    notif.showNotification({title: "RELOAD AIRCRAFT", theme: "GAMEPLAY", message:"Reload the aircraft to apply settings"});
+                };
+                popup.showPopUp({ title: "CRITICAL SETTING CHANGED", message: "Please reload the aircraft to apply your new settings.", style: "small"}, showNotif, showNotif);
```
Comment (Benjozork) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_MCDU_KB.js:40]: **thought:** The exact same code is repeated twice. Can we extract this to one function ?

**Review by Benjozork — CHANGES_REQUESTED** [review]

File: src/shared/src/notification.ts [hunk 7]
```diff
@@ -0,0 +1,105 @@
+/* eslint-disable camelcase */
+/* eslint-disable no-undef */
+/// <reference path="../../../typings/fs-base-ui/html_ui/JS/common.d.ts" />
```
Comment (Benjozork) [inline src/shared/src/notification.ts:3]: I think this is not needed anymore after the typings changes in #5359.

File: src/shared/src/popup.ts [hunk 8]
```diff
@@ -0,0 +1,97 @@
+/* eslint-disable no-undef */
+/* eslint-disable no-underscore-dangle */
+/// <reference path="../../../typings/fs-base-ui/html_ui/JS/common.d.ts" />
```
Comment (Benjozork) [inline src/shared/src/popup.ts:3]: I think this is not needed anymore after the typings changes in #5359.

Comment (2hwk) [reply src/shared/src/popup.ts]: image

Still needed

File: src/instruments/src/EFB/Settings/Settings.tsx [hunk 9]
```diff
@@ -31,6 +32,7 @@ const PlaneSettings = () => {
     const [accelerationOutAltSetting, setAccelerationOutAltSetting] = useState(accelerationOutAlt);
     const [defaultBaro, setDefaultBaro] = usePersistentProperty('CONFIG_INIT_BARO_UNIT', 'IN HG');
     const [weightUnit, setWeightUnit] = usePersistentProperty('CONFIG_USING_METRIC_UNIT', '1');
+    const notification = new Notification();
```
Comment (Benjozork) [inline src/instruments/src/EFB/Settings/Settings.tsx:35]: I'm not sure if instantiating those objects every render is a good idea.

Maybe store them in a `useState(() => new Notification())`, or use them at the call site (probably the latter).

**Review by Benjozork — APPROVED** [review]

File: src/instruments/src/EFB/Settings/Settings.tsx [hunk 10]
```diff
@@ -132,7 +133,10 @@ const AircraftConfigurationPage = () => {
                 <SelectGroup>
                     {weightUnitButtons.map((button) => (
                         <SelectItem
-                            onSelect={() => setWeightUnit(button.setting)}
+                            onSelect={() => {
+                                setWeightUnit(button.setting);
+                                new Notification().showNotification({ title: 'RELOAD AIRCRAFT', theme: 'GAMEPLAY', message: 'Reload the aircraft to apply settings.' });
```
Comment (tracernz) [inline src/instruments/src/EFB/Settings/Settings.tsx:138]: BTW, removing this warning (and the need for it when changing weight unit entirely in #5344).

**Comment by NUTTTTTTTT** [comment]
Quality Assurance Tester Report

Discord: █▀█ █▄█ ▀█▀#2123
Object of testing: #PR 5261
Tier of Testing: 1
Date: 18/08/2021

**Testing Process:**
Settings Reload:
- Loaded in on runway at EGCC
- Changed weights in the MCDU to LB's
- Observed pop up: came up as expected, shown in media. - PASS
- Cleared pop up, notification should show for 10 seconds. - FAIL - none showed.
- Reloaded sim and flight numerous times to repeat and retry throughout testing. 

Calibration Notification:
- Deleted ThrottleConfiguration.ini
- Loaded in from runway at EGCC again, departed for EDDB.
- Got to cruise at FL200 and CRZ page on lower ECAM.
- Made sure that AP is on and A/THRST is active.
- Set throttle a bit behind the CLB detent, so that LVR CLB flashed on the PFD.
- Left it in this state for 40 minutes total, did not move the throttles.

No notification was shown throughout the 40 minute cruise flight for this at all. - FAIL

Negatives: 
- Notifications not present at all.
- Pop up has two options for only one result.

Testing Results: FAIL

Conclusions: Pop up message appeared for settings reload, although notifications did not appear at all.  Options for yes and no given for a restart on the pop up, although both just close the prompt.

Media:
image

**Comment by FlightSimmer7700** [comment]
> 
> 
> Quality Assurance Tester Report
> 
> Discord: █▀█ █▄█ ▀█▀#2123
> Object of testing: #PR 5261
> Tier of Testing: 1
> Date: 18/08/2021
> 
> **Testing Process:**
> Settings Reload:
> 
>     * Loaded in on runway at EGCC
> 
>     * Changed weights in the MCDU to LB's
> 
>     * Observed pop up: came up as expected, shown in media. - PASS
> 
>     * Cleared pop up, notification should show for 10 seconds. - FAIL - none showed.
> 
>     * Reloaded sim and flight numerous times to repeat and retry throughout testing.
> 
> 
> Calibration Notification:
> 
>     * Deleted ThrottleConfiguration.ini
> 
>     * Loaded in from runway at EGCC again, departed for EDDB.
> 
>     * Got to cruise at FL200 and CRZ page on lower ECAM.
> 
>     * Made sure that AP is on and A/THRST is active.
> 
>     * Set throttle a bit behind the CLB detent, so that LVR CLB flashed on the PFD.
> 
>     * Left it in this state for 40 minutes total, did not move the throttles.
> 
> 
> No notification was shown throughout the 40 minute cruise flight for this at all. - FAIL
> 
> Negatives:
> 
>     * Notifications not present at all.
> 
>     * Pop up has two options for only one result.
> 
> 
> Testing Results: FAIL
> 
> Conclusions: Pop up message appeared for settings reload, although notifications did not appear at all. Options for yes and no given for a restart on the pop up, although both just close the prompt.
> 
> Media:
> image

Same negatives found on my qa test yesterday but did not report because we thought it was an installation issue with my sim

**Comment by 2hwk** [comment]
Need someone specifically with the Steam Version to test this to isolate the issue to the MS store version.

**Comment by FlightSimmer7700** [comment]
Quality Assurance Tester

Discord: FlightSimmer7700#3468
Object of testing: #5261 
Tier of Testing: 1
Date: 20/8/21

Testing Process: 
Loaded C&D and tried all options with reload needed
Loaded on runway and tested that popup did not appear for throttle calibration with calibrated throttles
Loaded mid-air with throttle calibration removed and tested for the popup (with a slightly edited version for lower times so not 10min)
Negatives: N/A
Testing Results: Passed

Conclusions: LGTM finally shows up after 10 tests on this pr 😝
image

**Comment by NUTTTTTTTT** [comment]
Quality Assurance Tester Report

Discord: █▀█ █▄█ ▀█▀#2123
Object of testing: #PR 5261
Tier of Testing: 1
Date: 20/08/2021

Testing Process:
Settings Reload:
- Loaded in on runway at EGSS
- Changed weights in the MCDU to LB's
- Observed pop up: came up as expected, shown in media. - PASS
- Cleared pop up, notification should show for 10 seconds. - PASS

Calibration Notification:
- Deleted ThrottleConfiguration.ini
- Loaded in from runway at EGSS again, departed runway heading.
- Got to cruise at FL150 and cruise page on ECAM.
- Made sure that AP is on and A/THRST is active.
- Set throttle a bit behind the CLB detent, so that LVR CLB flashed on the PFD.

After around 10 minutes the notification showed as expected, stuck for around 1 minute and then disappeared. Roughly 5 mins later the notification came back, lasted another minute. 

Negatives: N/A
Testing Results: PASS

Conclusions: Pop up message appeared for settings reload.. Options for yes and no given for a restart on the pop up, although both just close the prompt for now but all worked as intended. Great stuff!

Media:
image
image

**Pull request merged**